### PR TITLE
Server-side Client Certificate Authentication

### PR DESCRIPTION
**Pull request by lumberj**
Server-side Client Certificate Authentication

Uses https://github.com/ripple/passport-client-certificate

Working example:

Server
```diff
   _makeRouter () {
     const router = new Router()
-    router.get('/health', health.getResource)
+
+    router.get('/health',
+        passport.authenticate(['client-cert']),
+        health.getResource)

```

Client
```js
var fs = require('fs');
var https = require('https');

var options = {
  hostname: 'localhost',
  port: 3000,
  path: '/health',
  method: 'GET',
  key: fs.readFileSync('client1-key.pem'),
  cert: fs.readFileSync('client1-crt.pem'),
  ca: fs.readFileSync('ca-crt.pem') // Trust the test servers self-signed cert
};

var req = https.request(options, function(res) {
  res.on('data', function(data) {
    process.stdout.write(data);
  });
});
req.end();
```

```shell
$ LEDGER_TLS_KEY=server-key.pem \
  LEDGER_TLS_CERTIFICATE=server-crt.pem \
  LEDGER_TLS_CA=ca-crt.pem \
  LEDGER_PUBLIC_HTTPS=true LEDGER_DB_SYNC=1 \
  LEDGER_DB_URI=sqlite://:memory: \
  npm start
```

Certificates for this example can be found at:
https://github.com/ripple/passport-client-certificate/tree/master/test/data

### CODE REVIEW

File: migrations/20160127110439-account-fingerprint.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+'use strict';
+
+module.exports = {
+  up: function (queryInterface, Sequelize) {
+    queryInterface.addColumn('Accounts', 'fingerprint', Sequelize.STRING)
```
Comment (lumberj) [inline migrations/20160127110439-account-fingerprint.js]: I think I should be adding the index in the migration as well

File: src/services/config.js [hunk 2]
```diff
@@ -20,11 +20,6 @@ if (admin_pass) {
   }
 }
 
-localConfig.auth = {
```
Comment (lumberj) [inline src/services/config.js]: Moved to five-bells-shared

**Comment by lumberj** [comment]
cc: @justmoon and @sentientwaffle. At the moment this is failing because `passport-client-cert` is not yet included, as per the commit message. This is tested locally and everything works.

File: migrations/20160127110439-account-fingerprint.js (hunk 1 above)
Comment (lumberj) [inline migrations/20160127110439-account-fingerprint.js]: Actually, I should take this out as we've removed migrations

**Comment by clark800** [comment]
Can you rebase?

File: package.json [hunk 3]
```diff
@@ -31,7 +31,7 @@
     "co-request": "^1.0.0",
     "deep-diff": "^0.3.0",
     "five-bells-condition": "~2.0.0",
-    "five-bells-shared": "~9.1.2",
+    "five-bells-shared": "~9.2.1",
```
Comment (MatthewPhinney) [inline package.json]: This was also updated here: https://github.com/interledger/five-bells-ledger/commit/d682b3c2480a981939f46db0afb06218251a8d7e, which is probably causing the merge conflicts.

File: src/services/auth.js [hunk 4]
```diff
@@ -54,6 +55,22 @@ passport.use(new HTTPSignatureStrategy(
...
+passport.use(new ClientCertStrategy((certificate, done) => {
+  if (!config.getIn(['auth', 'client_certificates_enabled'])) {
+    return done(new UnauthorizedError('Unsupported authentication method'))
+  }
+
+  const fingerprint = certificate.fingerprint.toUpperCase()
+  Account.findByFingerprint(fingerprint)
+    .then(function (userObj) {
+      if (!userObj || userObj.is_disabled || !userObj.fingerprint ||
+          !fingerprint || userObj.fingerprint !== fingerprint) {
```
Comment (MatthewPhinney) [inline src/services/auth.js]: What's the rational behind checking `!fingerprint` here?

File: src/lib/app.js [hunk 5]
```diff
@@ -59,7 +60,33 @@ class App {
     yield this.db.init()
     yield seedDB(this.config)
 
-    this.koa.listen(this.config.getIn(['server', 'port']))
+    if (this.config.getIn(['server', 'secure'])) {
+      const https = require('https')
+      const tls = this.config.get('tls').toJS()
+
+      // https://nodejs.org/api/ssl.html#ssl_ssl_createserver_options_secureconnectionlistener
```
Comment (naoitoi) [inline src/lib/app.js]: I would elaborate what this URL is.

**Comment by naoitoi** [comment]
LGTM.

File: src/models/db/account.js [hunk 6]
```diff
@@ -54,6 +54,7 @@ class Account extends Model {
     delete data.primary
     delete data.password_hash
     delete data.public_key
+    delete data.fingerprint
```
Comment (clark800) [inline src/models/db/account.js]: What's the reason for deleting?

File: src/services/auth.js (hunk 4 above)
Comment (lumberj) [inline src/services/auth.js]: Good catch, should have deleted that

File: src/lib/app.js (hunk 5 above)
Comment (lumberj) [inline src/lib/app.js]: I think I'll remove it because the link can go stale or change

File: src/models/db/account.js (hunk 6 above)
Comment (lumberj) [inline src/models/db/account.js]: II think for the same reason the public_key is not shared

Comment (lumberj) [inline src/models/db/account.js]: The fingerprint is a hash of the contents of the client certificate. I don't think there's any case where this needs to be communicated back to a client

**Comment by MatthewPhinney** [comment]
LGTM

**Pull request merged**